design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,126 @@
+1. With the object store IP, port, and user name from step 1, the operator makes an S3 API call to create a new bucket with the CephObjectStoreUser as the owner.
+1. The operator creates a bucket ConfigMap in the dev-user namespace. The Host, Port, BucketName, and SSL (bool) are stored under data: .  For the purpose of bookkeeping, the namespace/CephObjectStoreUser.metadata.name and namespace/Secret.metadata.name are stored as annotations.
+1. An app Pod may then mount the Secret (annotated in the config map)  and the ConfigMap keys to begin accessing the bucket. 
+
+**Use Case:** Delete an Object Bucket 
+_As a Kubernetes user, I want to use kubectl to delete object bucket API resources owned by a given object user. (Note: the underlying S3 bucket is not deleted. Permanent deletion is left to the admin.)_
+
+1. The user deletes the CephObjectBucket.  This triggers an automated cleanup sequence:
+1. The CephObjectBucket is marked for deletion first and left in the foreground. Deletion is blocked by a finalizer on the CephObjectBucket, pending step (b).
+1. The respective ConfigMap is deleted.
+1. The finalizer in the CephObjectBucket is removed and the object is garbage collected.
+
+## API Specifications
+
+### CephObjectBucket
+
+```yaml
+apiVersion: object.k8s.io/v1
+kind: CephObjectBucket
+metadata:
+  name: my-bucket-1
+  namespace: dev-user
+  ownerReferences: # added by provisioner
+  - apiVersion: ceph.rook.io/v1
+    blockOwnerDeletion: true
+    kind: ObjectStoreUser
+    name: my-user-1
+    uid: 1234-3456-7839-1231
+spec:
+  bucketName:<string>         [1]
+  generateBucketName:<string> [1]
+  ownerUser: my-user # required, name of object user’s metadata.Name
+status:
+  phase: [ “creating”, “available”, ...]
+
+```
+
+[1] `bucketName` and `generateBucketName` are mutually exclusive. It is expected that `generateBucketName` be the common usage. This field produces a random bucket name beginning with the value of this stanza. If bucketName is defined then that value alone is used as the name of the bucket. Within an object store, buckets must be unique, so if bucketName is used there may be a higher chance of name collision. The api will continue retrying if the bucket cannot be created.
+
+### CephObjectBucketConfigMap
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: rook-ceph-object-bucket-my-bucket-1
+  namespace: dev-user
+  annotations:
+    bucket_secret: “rook-ceph-object-user-my-user-1”
+    bucket_owner: “my-user-1”
+  ownerReferences:
+  - name: my-bucket-1
+    uid: 1234-qwer-4321-rewq
+    apiVersion: object.k8s.io/v1
+    kind: CephObjectBucket 
+  finalizers:
```

Comment:
@travisn We've tried to make use of finalizers to make the system more robust against user error.  Do you think this is a better option than either:
1) let it be deleted, do nothing
2) let it be deleted but regenerate it

Is there a common practice for this in rook?

Reply:
In Rook we assume the user is an admin and they will have full access to the resources. If they delete resources rook relies on, it could kill the cluster. I don't recall hearing feedback from users that they think this is a problem.

However, in this case, it's not the admin who would be affected. Perhaps end users might be more prone to accidental operations? In my experience, finalizers should be used very sparingly. You have to rely on the rook operator to automate removal of the finalizer. If the operator is stopped before a finalizer is removed, the resource will live forever. As an example, we do have a finalizer on the CephCluster CRD and because users might stop the operator before removing the cluster, we end up with special [cleanup documentation](https://rook.io/docs/rook/v0.9/ceph-teardown.html#removing-the-cluster-crd-finalizer).

Reply:
@travisn For now I'm ok with this level of manual cleanup in cases where the operator is unavailable. At least the finalizer can be removed and the protected object deleted, but normally the operator would handle deletion of the finalizer.